mozilla_vpn/views/funnel_analysis.view.lkml

```diff
@@ -102,7 +102,7 @@ view: funnel_analysis {
 
 view: event_types {
   derived_table: {
-    sql: SELECT mozfun.event_analysis.aggregate_match_strings( ARRAY_AGG(CONCAT(COALESCE(mozfun.event_analysis.escape_metachars(property_value.value), ''),mozfun.event_analysis.event_index_to_match_string(et.index)))) AS match_string FROM `mozdata.mozilla_vpn.event_types` as et LEFT JOIN UNNEST(COALESCE(event_properties, [])) AS properties LEFT JOIN UNNEST(properties.value) AS property_value WHERE {% condition category %} category {% endcondition %} AND {% condition event %} event {% endcondition %} AND {% condition property_name %} properties.key {% endcondition %} AND {% condition property_value %} property_value.key {% endcondition %} ;;
+    sql: SELECT mozfun.event_analysis.aggregate_match_strings( ARRAY_AGG(CONCAT(COALESCE({% if property_value.value %} mozfun.event_analysis.escape_metachars(property_value.value), {% endif %} ''), mozfun.event_analysis.event_index_to_match_string(et.index)))) AS match_string FROM `mozdata.mozilla_vpn.event_types` as et LEFT JOIN UNNEST(COALESCE(event_properties, [])) AS properties LEFT JOIN UNNEST(properties.value) AS property_value WHERE {% condition category %} category {% endcondition %} AND {% condition event %} event {% endcondition %} AND {% condition property_name %} properties.key {% endcondition %} AND {% condition property_value %} property_value.key {% endcondition %} ;;
```

Comment:
- Are you sure `{% if property_value.value %}` will work?  It seems like maybe that's a reference to the SQL column, which won't be evaluatable in Liquid logic.
  - Something like `{% if _filters['property_value'] %}` might work ([the docs](https://docs.looker.com/reference/liquid-variables#_filters) say `_filters[...]` is allowed in the `sql` parameter of a derived table), though it depends on what `_filters[...]` evaluates to when the filter isn't set, since Liquid only considers boolean false values and null/nil values to be false (e.g. empty strings are not considered false).
  - There's also an [_is_filtered](https://docs.looker.com/reference/liquid-variables#is_filtered) property, though I think that might only work with view fields and not custom filters.
- This needs to preserve the existing behavior if _either_ of the `property_name` or `property_value` filters have been selected.
- With this approach of modifying the `ARRAY_AGG()` expression `ARRAY_AGG(DISTINCT ...)` should be used, otherwise it'll generate an unnecessarily repetitive regular expression, and we lose the benefit of having a much simpler regular expression when no property name/value filters are used.
  - Another possible approach is to dynamically add a false condition on the property values left join so nothing is joined if no property name/value filters are used:
    ```sql
    LEFT JOIN UNNEST(properties.value) AS property_value {% unless _filters['property_name'] or _filters['property_value'] %} ON FALSE {% endunless %}
    ```

Reply:
Hey @sean-rose ,

You're right. It didn't quite work as expected. I tried the `_filter` approach instead and we still get results and the SQL generation seems more like what I expected:
<img width="687" alt="image" src="https://user-images.githubusercontent.com/42538694/180203852-3d0f441a-eab9-415d-ab6c-dd72edf2c12c.png">

And we still get results for other steps now:
<img width="786" alt="image" src="https://user-images.githubusercontent.com/42538694/180203973-da511ec2-6a40-4ef1-b2eb-895682d77cba.png">

step_1 and step_2 correctly contain: ARRAY_AGG(CONCAT(COALESCE( `mozfun.event_analysis.escape_metachars(property_value.value),`  ''), mozfun.event_analysis.event_index_to_match_string(et.index))

and step_3 and step_4 exclude the function call since the filter for property_value is not set:
`ARRAY_AGG(CONCAT(COALESCE( ''), mozfun.event_analysis.event_index_to_match_string(et.index)))`

However, do you think this would be an appropriate solution?

Reply:
- It needs to also check for the `property_name` filter.
- It'd be best to use `ARRAY_AGG(DISTINCT ...)` with this approach to avoid having an unnecessarily repetitive regular expression.
- It'd be a bit cleaner to omit the `COALESCE('')` when no property name/value filters are specified (also in terms of how this would be implemented in [funnel_analysis_view.py](https://github.com/mozilla/lookml-generator/blob/4c72682d0288462e75d1f48ffc1c34a8222494a1/generator/views/funnel_analysis_view.py#L158-L173)).
- [Whitespace control](https://shopify.github.io/liquid/basics/whitespace/) could be used to avoid weird spaces in the generated SQL.

```suggestion
    sql: SELECT mozfun.event_analysis.aggregate_match_strings(ARRAY_AGG(DISTINCT CONCAT({% if _filters['property_name'] or _filters['property_value'] -%} COALESCE(mozfun.event_analysis.escape_metachars(property_value.value), ''), {% endif -%} mozfun.event_analysis.event_index_to_match_string(et.index)))) AS match_string FROM `mozdata.mozilla_vpn.event_types` as et LEFT JOIN UNNEST(COALESCE(event_properties, [])) AS properties LEFT JOIN UNNEST(properties.value) AS property_value WHERE {% condition category %} category {% endcondition %} AND {% condition event %} event {% endcondition %} AND {% condition property_name %} properties.key {% endcondition %} AND {% condition property_value %} property_value.key {% endcondition %} ;;
```

Or an alternative is to change the properties join (before I incorrectly said the property values join):
```suggestion
    sql: SELECT mozfun.event_analysis.aggregate_match_strings(ARRAY_AGG(CONCAT(COALESCE(mozfun.event_analysis.escape_metachars(property_value.value), ''), mozfun.event_analysis.event_index_to_match_string(et.index)))) AS match_string FROM `mozdata.mozilla_vpn.event_types` as et LEFT JOIN UNNEST(COALESCE(event_properties, [])) AS properties {% unless _filters['property_name'] or _filters['property_value'] -%} ON FALSE {% endunless -%} LEFT JOIN UNNEST(properties.value) AS property_value WHERE {% condition category %} category {% endcondition %} AND {% condition event %} event {% endcondition %} AND {% condition property_name %} properties.key {% endcondition %} AND {% condition property_value %} property_value.key {% endcondition %} ;;
```